### ISSUE TITLE
grad_clipping clips gate activations, not gradient

### ISSUE THREAD

**Issue by Richi91** [Lasagne/Lasagne#401]
I don't understand the way grad_clipping is implemented in the recurrent layers. Instead of clipping gradients, gate activations get clipped. Is there a reason for this? 1. I don't know any publication doing this and 2. it doesn't solve the exploding gradient problem.

The correct implementation should clip the gradient and it therefore should be implemented in the update functions.

Also you use theano.gradient.grad_clip, which works on each parameter independently (= hard clipping). Consider limiting the L2 norm of the gradient to a certain value, dividing each parameter by the same value and thus preserving the correct directiong of the gradient.
See [On the difficulty of training Recurrent Neural Networks](http://www.jmlr.org/proceedings/papers/v28/pascanu13.pdf), Algorithm 1. 

I use it the following way: 
(copied from https://github.com/kastnerkyle/minet/blob/master/minet/net.py#L236 )
```
def sgd_with_grad_clipping(loss_or_grads, params, learning_rate, rescale):
    grads = lasagne.updates.get_or_compute_grads(loss_or_grads, params)
    updates = OrderedDict()

    grad_norm = T.sqrt(sum(map(lambda x: T.sqr(x).sum(), grads)))
    not_finite = T.or_(T.isnan(grad_norm), T.isinf(grad_norm))
    grad_norm = T.sqrt(grad_norm)
    scaling_num = rescale
    scaling_den = T.maximum(rescale, grad_norm)
    for n, (param, grad) in enumerate(zip(params, grads)):
        grad = T.switch(not_finite, 0.1 * param,
            grad * (scaling_num / scaling_den))
        updates[param] = param - learning_rate * grad
    return updates   
```

**Comment by f0k** [comment]
You can actually just do:
```python
max_norm = 5.0
grads = theano.gradient(loss, params)
grads = [lasagne.updates.norm_constraint(grad, max_norm, range(grad.ndim))
         for grad in grads]
updates = lasagne.updates.whatever(grads, params)
```
Except that this doesn't have the finiteness check. In any case, it's easier to compute and clip the gradients and pass them to one of the update functions instead of writing your own update function.

About how it's implemented in the recurrent layers I'll leave it to @skaae or @craffel to reply!

**Comment by skaae** [comment]
We have reference in the docs. 
The grad_clip op clips the gradient in the backward pass and does nothing in the backward pass. 
From theanos description of gradclip:

          This op do a view in the forward, but clip the gradient.

          This is an elemwise operation.

You can also limit the L2 norm. I guess the choice is a hyperparameter. 

I have also found [T.clip(g) for g in T.grad(...)] to work well. 

best regards Søren

**Comment by craffel** [comment]
> The grad_clip op clips the gradient in the backward pass and does nothing in the ~~backward~~ forward pass. 

FTFY.

**Comment by skaae** [comment]
Thanks :)

**Comment by f0k** [comment]
> I have also found [T.clip(g) for g in T.grad(...)] to work well.

With the difference being that `grad_clip` in the recurrent layer will clip the gradient at every time step when backpropagating through time, not only in the end, right? (I would have expected this to matter a lot.)

**Comment by skaae** [comment]
Yes, thats my understanding
Den 01/09/2015 23.15 skrev "Jan Schlüter" <notifications@github.com>:

**Comment by Richi91** [comment]
Thanks for your answers!
I just experienced that the grad_clipping in the recurrent layer was not sufficient for preventing exploding/vanishing gradients. After very few iterations I had infs and I tested grad_clipping on a large range of values.

When limiting the L2 norm of the gradients (and not using grad_clipping of the recurrent layers), I had no exploding gradients.
But I agree that the choice can be seen as a hyperparameter.
